File: C++/M1files/lexer.cpp
#include <iostream>
#include <string>
#include <regex>

#include "HeaderFiles/tokens.h"
#include "HeaderFiles/lexer.h"

std::string _source;
char _character;
int _position;
int _read_position;

void _read_character() {
    if (_read_position >= _source.size()) {
        _character = '\0';
    } else {
        _character = _source[_read_position];
    }
    _position = _read_position;
    _read_position++;
}

char _peek_character() {
    if (_read_position >= _source.size()) {
        return '\0';
    }
    return _source[_read_position];
}

Token _make_two_character_token(TokenType token_type) {
    char prefix = _character;
    _read_character();
    char suffix = _character;
    return Token(token_type, std::string(1, prefix) + suffix);
}

void _skip_whitespace() {
    while (std::regex_match(std::string(1, _character), std::regex("^\\s$"))) {
        _read_character();
    }
}

bool _is_letter(char character) {
    return std::regex_match(std::string(1, character), std::regex("^[a-zA-ZáéíóúÁÉÍÓÚñÑ_]$"));
}

bool _is_number(char character) {
    return std::regex_match(std::string(1, character), std::regex("^\\d$"));
}

std::string _read_number() {
    int initial_position = _position;
    while (_is_number(_character)) {
        _read_character();
    }
    return _source.substr(initial_position, _position - initial_position);
}

std::string _read_identifier() {
    int initial_position = _position;
    while (_is_letter(_character)) {
        _read_character();
    }
    return _source.substr(initial_position, _position - initial_position);
}


/*Lexer(const std::string &source): _source(source), _character('\0'), _position(0), _read_position(0) {
    _read_character();
}*/

Token next_token() {
    _skip_whitespace();
    Token token(TokenType::ILLEGAL, std::string(1, _character)); // Default to ILLEGAL token

    if (_character == '=') { // Assign = or Equals ==
        if (_peek_character() == '=') {
            token = _make_two_character_token(TokenType::EQ);
        } else {
            token = Token(TokenType::ASSIGN, std::string(1, _character));
        }
    } else if (_character == '!') { // Negation ! or Dif !=
        if (_peek_character() == '=') {
            token = _make_two_character_token(TokenType::DIF);
        } else {
            token = Token(TokenType::NEGATION, std::string(1, _character));
        }
    } else if (_character == '+') { // Plus +
        token = Token(TokenType::PLUS, std::string(1, _character));
    } else if(_character == '{') { // Lbrace {
        token = Token(TokenType::LBRACE, std::string(1, _character));
    } else if(_character == '}') { // Rbrace }
        token = Token(TokenType::RBRACE, std::string(1, _character));
    } else if(_character == '(') { // Lparen (
        token = Token(TokenType::LPAREN, std::string(1, _character));
    } else if(_character == ')') { // Rparen )
        token = Token(TokenType::RPAREN, std::string(1, _character));
    } else if(_character == '-') { // Minus -
        token = Token(TokenType::MINUS, std::string(1, _character));
    } else if(_character == ',') { // Comma ,
        token = Token(TokenType::COMMA, std::string(1, _character));
    } else if(_character == ';') { // Semicolon ;
        token = Token(TokenType::SEMICOLON, std::string(1, _character));
    } else if (_is_letter(_character)) { // Letter
        std::string literal = _read_identifier();
        TokenType token_type = lookup_token_type(literal);
        return Token(token_type, literal);
    } else if (_is_number(_character)) { // Number
        std::string literal = _read_number();
        TokenType token_type = lookup_token_type(literal);
        return Token(TokenType::INTEGER, literal);
    } else if(_character == '>') { // Greater than > or equals >=
        if(_peek_character() == '=')
            token = _make_two_character_token(TokenType::GTE);
        else
            token = Token{TokenType::GT, std::string(1, _character)};
    } else if(_character == '<') // Less than < or equals <=
    {
        if(_peek_character() == '=')
            token = _make_two_character_token(TokenType::LTE);
        else
            token = Token{TokenType::LT, std::string(1, _character)};
    } else if(_character == '\0') { // End of file
        token = Token(TokenType::EOF_TOKEN, std::string(1, _character));
    }
    _read_character();
    return token;
}

File: Java/M1files/Lexer.java
package M1files;

public class Lexer {
    private String source;
    private String character;
    private int position;
    private int readPosition;

    public Lexer(String source) {
        this.source = source;
        this.position = 0;
        this.readPosition = 0;
        readCharacter();
    }

    public Token nextToken(){
        skipWhitespace();
        Token token = new Token(TokenType.ILLEGAL, character);
        if(character.equals("=")) {
            if(peekCharacter().equals("=")) { // Assing or equals '=' or '=='
                token =  makeTwoCharacterToken(TokenType.EQ); // '=='
            } else {
                token = new Token(TokenType.ASSIGN, character); // '='
            }
        } else if(character.equals(",")) { // Comma ,
            token = new Token(TokenType.COMMA, character);
        } else if(character.equals("!")) { // Different or equals '!='
            if(peekCharacter().equals("=")) { 
                token =  makeTwoCharacterToken(TokenType.DIF); // '!='
            } else {
                token =  new Token(TokenType.NEGATION, character); // '!'
            }
        } else if(character.equals("/")) { // Division
            token = new Token(TokenType.DIVISION, character);
        } else if(character.equals("")) { // Blank
            token = new Token(TokenType.EOF, character);
        } else if(Character.isLetter(character.charAt(0))) { // Letter
            token = new Token(Translator.lookupTokenType(readIdentifier()), readIdentifier());
        } else if(character.equals(">")) { // Greater than 
            if(peekCharacter().equals("=")) {
                token = makeTwoCharacterToken(TokenType.GTE); // '>='
            } else {
                token = new Token(TokenType.GT, character); // '>'
            }
        } else if(Character.isDigit(character.charAt(0))){ // Number
            token = new Token(Translator.lookupTokenType(readNumber()), readNumber());
        } else if(character.equals("{")) { // Rbrace '{'
            token = new Token(TokenType.LBRACE, character);
        } else if(character.equals("(")) { // Lparen '('
            token = new Token(TokenType.LPAREN, character);
        } else if(character.equals("<")) { // Less than
            if(peekCharacter().equals("=")) {
                token = makeTwoCharacterToken(TokenType.LTE); // '<='
            } else {
                token = new Token(TokenType.LT, character); // '<'
            }
        } else if(character.equals("-")) { // Minus '-'
            token = new Token(TokenType.MINUS, character);
        } else if(character.equals("*")) { // Multiplication '*'
            token = new Token(TokenType.MULTIPLICATION, character);
        } else if(character.equals("+")) { // Plus '+'
            token = new Token(TokenType.PLUS, character);
        } else if(character.equals("}")){ // Rbrace '}'
            token = new Token(TokenType.RBRACE, character);
        } else if(character.equals(")")){ // Rparen ')'
            token = new Token(TokenType.RPAREN, character);
        } else if(character.equals(";")) { // Semicolon ';'
            token = new Token(TokenType.SEMICOLON, character);
        }
        readCharacter();
        return token;
    }

    private String readNumber(){
        int initialPosition = position;
        while(/*position < source.length() &&*/ Character.isDigit(character.charAt(position))) {
            readCharacter();
        }
        return source.substring(initialPosition, position);
    }
    private String readIdentifier(){
        int startPosition = position;
        while(position < source.length() && Character.isLetter(character.charAt(position))) {
            readCharacter();
        }
        return source.substring(startPosition, position);
    }
    private void readCharacter(){
        /*if(readPosition >= source.length()) {
            character = "";
        } else {
            character = source.substring(readPosition, readPosition + 1);
        }*/
        if (readPosition < source.length()) {
            character = source.substring(readPosition, readPosition + 1);
            position = readPosition;
            readPosition++;
        } else {
            character = "";
        }
    }
    private String peekCharacter(){
        if(readPosition >= source.length()) {
            return "";
        } 
        return source.substring(readPosition, readPosition + 1);
    }
    private Token makeTwoCharacterToken(TokenType tokenType){
        String prefix = character;
        readCharacter();
        String suffix = character;
        return new Token(tokenType, prefix + suffix);
    }
    private void skipWhitespace(){
        while(character.equals(" ") || character.equals("\t") || character.equals("\n") || character.equals("\r")) {
            readCharacter();
        }
    }
}